lib/smack/VerifierCodeMetadata.cpp

```diff
@@ -0,0 +1,124 @@
+//
+// This file is distributed under the MIT License. See LICENSE for details.
+//
+
+#define DEBUG_TYPE "contracts"
+
+#include "smack/SmackOptions.h"
+#include "smack/Naming.h"
+#include "smack/VerifierCodeMetadata.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/IR/DataLayout.h"
+#include "llvm/IR/Dominators.h"
+
+#include <vector>
+#include <stack>
+#include <map>
+#include <set>
+
+namespace smack {
+
+using namespace llvm;
+
+namespace {
+  std::string getString(const llvm::Value* v) {
+    if (const llvm::ConstantExpr* constantExpr = llvm::dyn_cast<const llvm::ConstantExpr>(v))
+      if (constantExpr->getOpcode() == llvm::Instruction::GetElementPtr)
+        if (const llvm::GlobalValue* cc = llvm::dyn_cast<const llvm::GlobalValue>(constantExpr->getOperand(0)))
+          if (const llvm::ConstantDataSequential* cds = llvm::dyn_cast<const llvm::ConstantDataSequential>(cc->getOperand(0)))
+            return cds ->getAsCString();
+    return "";
+  }
+
+  void mark(Instruction &I, bool V = true) {
+    auto& C = I.getContext();
+    I.setMetadata("verifier.code",
+      MDNode::get(C, ConstantAsMetadata::get(
+        V ? ConstantInt::getTrue(C) : ConstantInt::getFalse(C)
+      )));
+  }
+
+  bool isMarked(Instruction& I) {
+    auto *N = I.getMetadata("verifier.code");
+    assert(N && "expected metadata");
+    assert(N->getNumOperands() == 1);
+    auto *M = dyn_cast<ConstantAsMetadata>(N->getOperand(0).get());
+    assert(M && "expected constant-valued metadata");
+    auto *C = dyn_cast<ConstantInt>(M->getValue());
+    assert(C && "expected constant-int-valued metadata");
+    return C->isOne();
+  }
+}
+
+void VerifierCodeMetadata::getAnalysisUsage(AnalysisUsage &AU) const {
+  AU.addRequired<DominatorTreeWrapperPass>();
+}
+
+bool VerifierCodeMetadata::runOnModule(Module &M) {
+  visit(M);
+  while (!workList.empty()) {
```

Comment:
It would take me some time to check this logic, but I trust  that you did it correctly. I would guess that you are being conservative if you encounter something like a store instruction - since you are not tracking uses through memory.
Also, and this occurred to me while back, it might be useful to actually add comments here and there in the code we write :).